controllers/rental.controller.js

```diff
@@ -0,0 +1,178 @@
+  }
+
+  /**
+   * @desc Create rental
+   * @route POST /api/v1/rentals
+   * @access Private
+   */
+  checkOut = asyncHandler(async (req, res) => {
+    // Add user to req.body
+    req.body.authorInformation = req.user;
+    const rawData = req.body;
+    await this.rentalLib.checkResourceInDB({ user: rawData.user, book: rawData.book });
+    const rental = await this.rentalLib.createRental(rawData);
+    return res.status(201).json({
+      success: true,
+      data: rental,
+    });
+  });
+
+  /**
+   * @desc Edit rental
+   * @route PUT /api/v1/rentals/:id
+   * @access Private
+   */
+  putRental = asyncHandler(async (req, res, next) => {
+    // Add user to req.body
+    const { id } = req.params;
+    req.body.authorInformation = req.user;
+    const rawData = req.body;
+
+    let rental = await this.rentalLib.fetchRental({ _id: id });
+    if (!rental) {
+      return next(
+        new ErrorResponse(`Rental with id: ${id} does not exist on the database`, 404),
+      );
+    }
+
+    rental = await this.rentalLib.updateRental(id, rawData);
+    return res.status(202).json({
+      success: true,
+      data: rental,
+    });
+  });
+
+  /**
+   * @desc Edit rental to add a like
+   * @route PUT /api/v1/rentals/:id/checkin
+   * @access Private
+   */
+  checkIn = asyncHandler(async (req, res, next) => {
+    // Add user to req.body
+    const { user, params } = req;
+    const { id } = params;
+
+    let rental = await this.rentalLib.fetchRental({ _id: id, user: user.id });
+    if (!rental) {
+      return next(
+        new ErrorResponse(`Rental with id: ${id} does not exist on the database`, 404),
+      );
+    }
+
+    rental = await this.rentalLib.updateRental(id, { isReturned: true });
+    return res.status(202).json({
+      success: true,
+      data: rental,
+    });
+  });
+
+  /**
+   * @desc Get rentals
+   * @route GET /api/v1/rentals
+   * @access Private
+   */
+  getRentals = asyncHandler(async (req, res) => {
+    const { query, params } = req;
+    const {
+      page, limit, select, sort, ...filter
+    } = query;
+    let localFilter = { ...filter };
+    if (params.bookId) {
+      localFilter = { ...filter, book: params.bookId };
+    }
+    const result = await advancedResults(Rental, localFilter, {
+      page: page || parseInt(page, 10),
+      limit: limit || parseInt(limit, 10),
+      select,
+      sort,
+    });
+
+    res.status(200).json({
+      success: true,
+      ...result,
+    });
+  });
+
+  /**
+   * @desc Get rental likes
+   * @route GET /api/v1/rentals/books/:userId
+   * @access Private
+   */
+  getBooksRented = asyncHandler(async (req, res) => {
+    // Add user to req.body
+    const { userId } = req.params;
+    req.body.authorInformation = req.user;
+
+    const rentals = await advancedResults(Rental, { user: userId }, {
+      page: 1,
+      distinct: 'book',
+      populate: 'book',
+    });
+
+    const books = await this.bookLib.fetchBooks({ _id: { $in: rentals.data } });
+
```

Comment:
Implementing it this way is fine but you may find it difficult to paginate(which is very likely to be needed as more books get checkout by this user) accurately. So you can achieve the same result with one db query with pagination as well.

Try this
```javascript
const rentals = await advancedResults(Rental, { user: userId }, {
      page: req.query.page,
      distinct: 'book',
      populate: 'book',
      select:'book'
    });
```
Or
```javascript
const rentals = await advancedResults(Rental, { user: userId }, {
      page: req.query.page,
      distinct: 'book',
      populate: 'book',
      select:'book._id, book.title, etc'
    });
```
Then 
```javascript
const books = {...rentals, data:rentals.data.map(r => r.book)}
```